### PR TITLE
Resolve preserve locality outside of direct keymap bindings

### PR DESCRIPTION
**Pull request by tokazio**
Seems to resolve https://github.com/zmkfirmware/zmk/issues/1494, at least resolve https://github.com/zmkfirmware/zmk/issues/1347

### CODE REVIEW

File: app/include/zmk/split/bluetooth/central.h [hunk 1]
```diff
@@ -5,4 +5,6 @@
 #include <zmk/behavior.h>
 
 int zmk_split_bt_invoke_behavior(uint8_t source, struct zmk_behavior_binding *binding,
-                                 struct zmk_behavior_binding_event event, bool state);
\ No newline at end of file
+                                 struct zmk_behavior_binding_event event, bool state);
+
+int zmk_run_behavior(struct zmk_behavior_binding *binding, struct zmk_behavior_binding_event event,uint8_t source,bool pressed);
```
Comment (petejohanson) [inline app/include/zmk/split/bluetooth/central.h:10]: I think this general functionality is useful, but a few things:

* This should be in `keymap.h`, not this header.
* I don't love the name. We are triggering the pressed/released callbacks, not "running" something.

File: app/src/keymap.c [hunk 2]
```diff
@@ -179,17 +179,23 @@ int zmk_keymap_apply_position_state(uint8_t source, int layer, uint32_t position
         .timestamp = timestamp,
     };
 
-    LOG_DBG("layer: %d position: %d, binding name: %s", layer, position,
-            log_strdup(binding.behavior_dev));
+    return zmk_run_behavior(&binding,event,source,pressed);
+}
+
+int zmk_run_behavior(struct zmk_behavior_binding *binding, struct zmk_behavior_binding_event event,uint8_t source,bool pressed){
```
Comment (petejohanson) [inline app/src/keymap.c:185]: `binding` probably should be `const` here. Also, be sure to run `clang-format` on any changed files for consistent formatting.

**Review by petejohanson — CHANGES_REQUESTED** [review]
Thanks for working on this! I think this is a good start, but some tweaks needed.

**Comment by tokazio** [comment]
Thanks for the review.

I see some delay when the trigger comes from the left part of the keyboard that i doesn't see when the trigger comes from the right part.

https://user-images.githubusercontent.com/7676865/214581300-7e72af8d-69ae-4c33-8654-1b414f0c1758.mov

**Comment by xudongzheng** [comment]
> I see some delay when the trigger comes from the left part of the keyboard that i doesn't see when the trigger comes from the right part.

This delay is expected. Since the connection interval is 7.5ms and connection latency is 30 (see https://github.com/zmkfirmware/zmk/blob/main/app/src/split/bluetooth/central.c#L412), it can take up to 7.5ms*(30+1)=232.5ms for the peripheral to receive the update from the central.

**Comment by petejohanson** [comment]
Exactly right... We could lower that connection latency, at the expense of increased power consumption on the peripheral side. (side note: it'd be interesting to renegotiate the connection parameters if the power status of the peripheral changes thanks to being plugged in, to improve responsiveness of this exact scenario....)

**Comment by vladsolokha** [comment]
> Thanks for the review.
> 
> I see some delay when the trigger comes from the left part of the keyboard that i doesn't see when the trigger comes from the right part.

Can you share how you make that behavior happen in your keymap? The layer switch and underglow change? Is there a way I  could do the same and have the rgb colors go back to the original effect when the layer is then switched off or released?

I have some rgb mods on &to layer switches in my keymap here, but I was hoping to do the same for my two main layers 1 and 2. I was hoping to leave them as sticky on tap and momentary on hold with the color changes. 

https://github.com/vladsolokha/zmk-config/blob/main/config/splitkb_aurora_sweep.keymap

**Comment by tokazio** [comment]
You Can found m'y config here https://github.com/tokazio/zmk-config-sofle 
Nothing special

**Comment by tokazio** [comment]
> > I see some delay when the trigger comes from the left part of the keyboard that i doesn't see when the trigger comes from the right part.
> 
> This delay is expected. Since the connection interval is 7.5ms and connection latency is 30, it can take up to 7.5ms*(30+1)=232.5ms for the peripheral to receive the update from the central.
> 
> https://github.com/zmkfirmware/zmk/blob/3d2bd017472c57f4b3860033f989a44f094febaa/app/src/split/bluetooth/central.c#L412

Oh ok, good to know it

**Comment by xudongzheng** [comment]
> Is there a possibilité to force the update when this précise Key is down ?

One option is defining a read service/characteristic on the central that you can read from the peripheral. This should have the current LED state.

After a key event such as a press/release on the peripheral happens, you read the state on central from the peripheral (rather than waiting for the central to write to peripheral). You can try to do this immediately or after a few milliseconds to give the central to process and update the state.

I'm not familiar enough with the Bluetooth specification to be certain that this wouldn't be subject to the latency value, but this is the first thing I would try to reduce the latency.

It's not trivial to implement and I would consider just reducing the latency to a lower value at the cost of some battery life.

**Comment by caksoylar** [comment]
@tokazio in case you missed it, @petejohanson noted a couple items here that needed changing: https://github.com/zmkfirmware/zmk/pull/1630#pullrequestreview-1262910398

**Comment by tokazio** [comment]
After some time of use i'm observing that colors, when i push down the upper/lower keys are not the ones i've programmed. Even the effect have changed. I have now the multicolors wave effect. Maybe there is a conflict with the 'persistence' feature of effect/color that is triggered after a given time

**Comment by seansfkelley** [comment]
Thanks for opening this; this resolved an issue I was having with underglow on the peripheral side. I also see noticeable lag between the central and peripheral sides, but for my usages (which are infrequent), that's fine.

@petejohanson it looks like your comments are only stylistic. Would you accept a PR with the same behaviors and your style points addressed, or would you rather prioritize something like https://github.com/zmkfirmware/zmk/pull/2036 instead? (It claims to resolve this issue as well, though I have not tried it.) I can open a cleaned-up PR if so.

**Comment by petejohanson** [comment]
> Thanks for opening this; this resolved an issue I was having with underglow on the peripheral side. I also see noticeable lag between the central and peripheral sides, but for my usages (which are infrequent), that's fine.
> 
> @petejohanson it looks like your comments are only stylistic. Would you accept a PR with the same behaviors and your style points addressed, or would you rather prioritize something like #2036 instead? (It claims to resolve this issue as well, though I have not tried it.) I can open a cleaned-up PR if so.

I'd rather we clean this up and get it merged, and then pursue #2036 as a next step.

**Comment by seansfkelley** [comment]
Sounds good, thanks. It seems like #2409 is currently active and does just that, so I won't add more noisy PRs just yet, but if that stalls out I'll open another PR like this but with focusing on resolving just the stylistic comments.

**Comment by caksoylar** [comment]
My understanding is that this PR is not actually solving the issue besides modifying only macros to invoke behaviors globally. #2409 should solve both global and source locality invocations for every kind of behavior that invokes other behaviors (currently with the exception of sensors not respecting source locality). As such this PR shouldn't be accepted even without the stylistic issues and I don't intend to stall #2409.

**Comment by caksoylar** [comment]
Closing as it is incorporated into #2409.

**Pull request closed without merging**